

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Hello World: What’s the purpose of bioinformatics? &mdash; ubic_crash_course 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The Linux Terminal" href="2_LinuxTerminal.html" />
    <link rel="prev" title="Welcome to the Bioinformatics Crash Course!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> ubic_crash_course
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Lessons</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Hello World: What’s the purpose of bioinformatics?</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#welcome-to-the-crash-course">Welcome to the Crash Course!</a></li>
<li class="toctree-l2"><a class="reference internal" href="#big-bioinformatics-problems">Big Bioinformatics Problems</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#i-the-alignment-problem-full-lesson">I. The Alignment Problem (full lesson)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ii-the-clustering-problem-full-lesson">II. The Clustering Problem (full lesson)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#some-types-of-clustering">Some types of clustering:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#your-turn-aliview">[Your Turn!] Aliview</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#prerequisites">Prerequisites</a></li>
<li class="toctree-l3"><a class="reference internal" href="#analysis">Analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#the-linux-terminal">The Linux Terminal</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="2_LinuxTerminal.html">The Linux Terminal</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_AdvancedTerminal.html">Advanced Command Line + Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Python.html">Python Basics</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_Biopython.html">Bioinformatics x Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Challenge1.html">Challenge 1</a></li>
<li class="toctree-l1"><a class="reference internal" href="8.1_Alignment.html">Alignment (Part I)</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_1_Clustering.html">Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_Phylogenetics.html">Phylogenetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="11_Challenge2.html">Challenge 2</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ubic_crash_course</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Hello World: What’s the purpose of bioinformatics?</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/1_Welcome.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="hello-world-what-s-the-purpose-of-bioinformatics">
<h1>Hello World: What’s the purpose of bioinformatics?<a class="headerlink" href="#hello-world-what-s-the-purpose-of-bioinformatics" title="Permalink to this headline">¶</a></h1>
<div class="section" id="welcome-to-the-crash-course">
<h2>Welcome to the Crash Course!<a class="headerlink" href="#welcome-to-the-crash-course" title="Permalink to this headline">¶</a></h2>
<p>These activities are designed for students with zero to some programming experience and zero to some bioinformatics experience.  The crash course is always in development, so we ask for everyone to be patient and tell us when we are screwing up. Thanks!</p>
</div>
<div class="section" id="big-bioinformatics-problems">
<h2>Big Bioinformatics Problems<a class="headerlink" href="#big-bioinformatics-problems" title="Permalink to this headline">¶</a></h2>
<p>As you progress through bioinformatics, you might find yourself developing technical skills without knowing what problems to tackle. In fact, there are several common and difficult problems in bioinformatics worth knowing. Therefore, this lesson will introduce you to of some of the things we, as bioinformaticians, do.</p>
<p><em>Note: In this course, we try to avoid lecture-based material. However, in this first lesson, we want to set a foundation of major problems bioinformaticians work on. We’ll keep it short, and move on to the interactive part: working in a Linux environment!</em></p>
<div class="section" id="i-the-alignment-problem-full-lesson">
<h3>I. The Alignment Problem (<a class="reference external" href="https://github.com/sabeelmansuri/Bioinformatics-Crash-Course/blob/master/5_Alignment">full lesson</a>)<a class="headerlink" href="#i-the-alignment-problem-full-lesson" title="Permalink to this headline">¶</a></h3>
<p>The first problem we’ll look at is alignment. <strong>Alignment is the process of arranging sequences in a way to identify regions of similarity that may be a consequence of functional, structural, or evolutionary relationships between the sequences.</strong> In order to understand why alignment is hard and practical problem, we will start by learning a bit about sequencing DNA.</p>
<p><strong>Illumina Sequencing:</strong></p>
<p>This is the most widely used “golden standard” method of sequencing DNA. The main idea is that DNA is fragmented into pieces, attached to a flow cell, copied many times over by PCR, and the complement strand is determined one nucleotide at a time. The modified nucleic acids Illumina uses during the generation of complementary strands emit light when they are bound. Illumina sequencing adds one base, measures the color output of the flow cell, adds a different base, measures the color output and repeats over and over. Here is a nice graphic depicting the process:</p>
<p><img alt="graphic explaining illumina" src="http://www.3402bioinformaticsgroup.com/wp-content/uploads/2016/07/NGS.png" /></p>
<p>There is also a video explaining Illumina sequencing which you’ll watch in <em>every single Bioinformatics class</em> you take in the future, found <a class="reference external" href="https://www.youtube.com/watch?v=fCd6B5HRaZ8">here</a>.</p>
<p>Want to learn more about different sequencing methods? We have a document describing non-Illumina methods in more detail <a class="reference internal" href="extras/1.1_Extra_Sequencing.html"><span class="doc">over here</span></a></p>
<p><em><strong>Comparing Sequencing Methods and Why they Matter</strong></em></p>
<p>What do all sequencing method have in common (yes, even PacBio)?  All produce ridiculous quantities of small DNA sequences. Illumina produces  300 million to 4 billion reads per run, with a selection of read lengths ranging from 50 base pairs to 300 base pairs. Meanwhile, Sanger produces 50000 sequences at lengths varying from 800 to 1000 base pairs. To give some perspective, the typical animal of interest is a human and those have 3.0×10^9 base pairs. Individual human genes range from 1148 to 37.7 kb (average length = 8446 bp,s.d. = 7124).</p>
<p>These tiny reads overlap all over the place. If you imagine the true sequence these reads came from and place the reads where they came from, you will get many reads piled up over every base pair in the true sequence. The more reads pile up, the more accurately you can predict the actual sequence. A common measure that rates the robustness of an alignment is coverage:</p>
<p><img alt="coverage" src="https://slideplayer.com/slide/5083621/16/images/4/Definition+of+Coverage.jpg" /></p>
<p>Besides helping us reconstruct the DNA put into a sequencer, alignment can also</p>
<ul class="simple">
<li><p>Determine which organism an unkown sequence comes from</p></li>
<li><p>Pinpoint locations where mutations have occured relative to a reference sequence</p></li>
<li><p>Help determine the evolutionary distance between sequences</p></li>
</ul>
<p>TLDR: There are numerous complex applications of bioinformatics algorithms, from functional structure predictions to ancestral reconstructions. Alignment serves as the foundation for many of these algorithms, making basic sense of the incomprehensible mass of DNA that sequencing gives us.</p>
</div>
<div class="section" id="ii-the-clustering-problem-full-lesson">
<h3>II. The Clustering Problem (<a class="reference external" href="https://github.com/sabeelmansuri/Bioinformatics-Crash-Course/blob/master/6.%20Clustering">full lesson</a>)<a class="headerlink" href="#ii-the-clustering-problem-full-lesson" title="Permalink to this headline">¶</a></h3>
<p><strong>Clustering is the the process of assigning data points to groups in such a way that the elements in a group/cluster are more similar to each other than they are to those in other groups.</strong> The definition of what it means to be similar can vary and is determined by the function we use to measure distance between two points.</p>
<div class="section" id="some-types-of-clustering">
<h4>Some types of clustering:<a class="headerlink" href="#some-types-of-clustering" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Hierarchical Clustering: Repeatedly combines the closest points into a cluster that is the hybrid location of both points. The reason this is interesting to us is that it forms a tree of clusters, which can represent a tree of related genes which can be used to infer homology.</p></li>
</ul>
<p><img alt="hierarchical cluster" src="https://upload.wikimedia.org/wikipedia/commons/a/ad/Hierarchical_clustering_simple_diagram.svg" /></p>
<ul class="simple">
<li><p>K-means Clustering: Separates a dataset into k groups of points in such a way that the members of a cluster are as close as possible to the center of the cluster they belong to. This type of clustering can check that the datapoints we are observing cluster together by tissue type, experimental conditions, time points, etc.</p></li>
<li><p>Fuzzy Clustering: Datapoints are not definitively assigned to a specific cluster, rather they are given a likelihood of belonging to a cluster. This can be used to ascertain levels of co expression between genes, revealing genes which may be under common regulatory control.</p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="your-turn-aliview">
<h2>[Your Turn!] Aliview<a class="headerlink" href="#your-turn-aliview" title="Permalink to this headline">¶</a></h2>
<p>Alright, enough of the conceptual stuff. Let’s get our hands dirty.</p>
<p><em>Note: The goal of this exercise is not really to master alignment. Instead, it’s more to get familiar with using bioinformatics tools and working with bioinformatics data.</em></p>
<div class="section" id="prerequisites">
<h3>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this headline">¶</a></h3>
<p>You will need Java to proceed. If you do not have it, go <a class="reference external" href="https://www.java.com/en/">here</a> and press the big red install button in the middle of the page.</p>
<p>Aliview is a sequence viewer with a bunch of built-in tools, including alignment tools. We will use Aliview to see what a typical dataset looks like coming out of the illumina sequencing machine and what it means, visually, to align the sequences. Click <a class="reference external" href="http://www.ormbunkar.se/aliview/">here</a> and go to download the stable version for your OS.</p>
</div>
<div class="section" id="analysis">
<h3>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h3>
<p>Next, download <a class="reference external" href="https://drive.google.com/a/ucsd.edu/file/d/1iIzDwKm2k_VOnen0BRwtZ9Jlf81h3t5g/view?usp=sharing">this neat dataset</a>. Launch aliview, click file-&gt;open file-&gt;PC64_V48_small_reconstructed_seqs.fasta. Scroll to the right and notice the mess that begins to form as you scroll.</p>
<p>These sequences are sourced from the same gene and have gone through many cleaning steps, so the differences between them are quite likely to be real (rather than artifacts from sequencing). Click align in the upper left corner and click realign everything. Now scroll forward and observe the gaps that have been inserted by the aligner.</p>
<p>These sequences are actually from HIV-1 glycoprotein envelopes from a person with broadly neutralizing antibodies against HIV-1(sequenced with PacBio). An effective HIV-1 vaccine should evoke the production of broadly neutralizing antibodies, so it is important to study the structure of the envelope that caused these antibodies to develop in individuals. The steps leading up to the data you downloaded allowed us to reveal a few specific strains. Now that we have them aligned, we can start asking questions about the differences between them and their evolution(which is important to figuring out how to counteract them). If you want to go into the nitty-gritty biology behind these sequences, go <a class="reference external" href="https://www.cell.com/immunity/pdf/S1074-7613(17)30479-X.pdf">here</a>. I will make a reconstruction of the evolutionary relationship for you to look at.</p>
<p>The SNP differences are obvious, but you will notice that there are weirder differences - like the &gt;100 bp gaps formed in the middle. The truth is that I do not know why those are there! Those can be</p>
<ol class="simple">
<li><p>Real differences in hypervariable loops</p></li>
<li><p>non-functional virions</p></li>
<li><p>RT/PCR artifacts</p></li>
</ol>
<p>Here’s another cool tool: Blast will quickly look up a sequence in the NCBI database and spit out similar sequences it finds. Now, go to aliview and click edit-&gt;delete all gaps in all sequences. Copy the first sequence into the clipboard. Google NCBI Blast and open up the first result. Paste the sequence into the big box in the top of the page and click <strong>BLAST</strong> at the bottom. After a few seconds, Blast should link you to a whole lot of glycoprotein that are… from the article these were published in! This is one of the basic uses of Blast - to figure out where a sequence comes from.</p>
</div>
</div>
<div class="section" id="the-linux-terminal">
<h2>The Linux Terminal<a class="headerlink" href="#the-linux-terminal" title="Permalink to this headline">¶</a></h2>
<p>Click <a class="reference internal" href="2_LinuxTerminal.html"><span class="doc">here</span></a> to move on to the next part of the lesson.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="2_LinuxTerminal.html" class="btn btn-neutral float-right" title="The Linux Terminal" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to the Bioinformatics Crash Course!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>