

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Codon Alignment Challenge &mdash; ubic_crash_course 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> ubic_crash_course
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Lessons</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="1_Welcome.html">Hello World: What’s the purpose of bioinformatics?</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_LinuxTerminal.html">The Linux Terminal</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_AdvancedTerminal.html">Advanced Command Line + Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Python.html">Python Basics</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_Biopython.html">Bioinformatics x Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Challenge1.html">Challenge 1</a></li>
<li class="toctree-l1"><a class="reference internal" href="8.1_Alignment.html">Alignment (Part I)</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_1_Clustering.html">Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_Phylogenetics.html">Phylogenetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="11_Challenge2.html">Challenge 2</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ubic_crash_course</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Codon Alignment Challenge</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/8.2_Codon_Alignment.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="codon-alignment-challenge">
<h1>Codon Alignment Challenge<a class="headerlink" href="#codon-alignment-challenge" title="Permalink to this headline">¶</a></h1>
<p>One thing biologists care a lot about is the way amino acids change through time. This is found by sequencing DNA at several timepoints, codon aligning various timepoints, and comparing timepoints to see which amino acids change at what time. Why don’t we just convert to amino acids and align there? Because different reading frames and deletions cause suboptimal translations into the amino acid realm, and therefore decrease the usefulness of the alignment. Plus, codon alignment can reveal synonymous vs nonsynonymous mutations.</p>
<p>For a visualization of this type of alignment, <a class="reference external" href="https://flea.ki.murrell.group/view/P018/sequences/">head over here</a>. The sequences shown here are from an HIV envelope. If you click on on amino acid index, you can see a graph showing how the amino acid in that position evolved over time. You can explore the site - I recommend the tree section because it is pretty.</p>
<p>I made a fake fasta of sequences that need to be codon aligned over at <code class="docutils literal notranslate"><span class="pre">/srv/WI20/not_aligned.fasta</span></code>. Here is my step by step guide on making your own codon aligner to make sure those sequences are nice and neat.</p>
<div class="section" id="get-the-imports">
<h2>0. Get the imports<a class="headerlink" href="#get-the-imports" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#to call mafft</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="c1">#to use the Seq type</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="c1">#to read/write fasta</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="c1">#to translate</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_dna</span>
</pre></div>
</div>
</div>
<div class="section" id="align-with-mafft-and-replace-initial-gaps-with-n-s">
<h2>1. Align with mafft and replace initial gaps with N’s<a class="headerlink" href="#align-with-mafft-and-replace-initial-gaps-with-n-s" title="Permalink to this headline">¶</a></h2>
<div class="section" id="why-we-do-this">
<h3>Why we do this?<a class="headerlink" href="#why-we-do-this" title="Permalink to this headline">¶</a></h3>
<p>Due to DNA fragmentation, sequencing starts at many different points, and we don’t know ahead of time where the points are. If we don’t identify where one sequence starts relative to another, we cannot begin comparing them. Why the N’s? That’s because the biological meaning of N’s is different than that of gaps. Gaps in an alignment indicate deletions or insertions from one sequence to another, we predict that something was removed or added. In the case of different starting points, we know that <em>something</em> is supposed to be there, since we have information from other sequences. Thus it is not an insertion or deletion, but unkown nucleotides. We represent this with N’s.</p>
<p><strong>a.</strong> Align the file with mafft. Syntax for this is provided in <a class="reference external" href="https://github.com/ubicucsd/bootcamp_19_20/blob/master/8_Alignment1">lesson 8</a></p>
<p><strong>b.</strong> Go through each of the sequences in the mafft aligned file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;nuc_aligned.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
</pre></div>
</div>
<p><strong>c.</strong> Count the number of initial gaps on each sequence</p>
<p><strong>d.</strong> Degap each sequence and place it into a Seq object</p>
<p>The syntax for this step is whack, so let me give it to you:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sequence</span><span class="o">=</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p><strong>d.</strong> Prepend each sequence with n’s. The number of n’s should be equal to the number of initial gaps that sequence had.</p>
</div>
</div>
<div class="section" id="correct-the-reading-frame-and-length">
<h2><strong>2.</strong> Correct the reading frame and length<a class="headerlink" href="#correct-the-reading-frame-and-length" title="Permalink to this headline">¶</a></h2>
<p>We have now lined up sequences with different starting points, but what if every single sequence starts at the wrong reading frame? All of the translated animo acids will be useless. One way of making the choice between starting each sequence from the first, second, or third nucleotide by seeing which one results in the longest total distance between stop codons in all of the sequences.</p>
<p>In the interest of time, let us assume that the reading frame is correct. All you need to do is to make sure that the alignment has a length that is a multiple of three.</p>
<p><strong>a.</strong> Go through each index from 0 to the number of sequences</p>
<p><strong>b.</strong> Use the modulus to find whether the current sequences has a length divisible by three.</p>
<p>What is a modulus? it’s this <code class="docutils literal notranslate"><span class="pre">%</span></code> symbol. In math, this symbol will find the remainder for you. So 15%3=0, 16%3=1, 17%3=2, 18%3=0 and so on.</p>
<p><strong>c</strong> Based on the remainder you found in b, figure out how much of the sequence you should cut off the end. Syntax hint: <code class="docutils literal notranslate"><span class="pre">seq[0:-1]</span></code> will give you the sequence with the last nucleotide cut off.</p>
</div>
<div class="section" id="translate-the-dna-into-amino-acids">
<h2><strong>3.</strong> Translate the DNA into amino acids.<a class="headerlink" href="#translate-the-dna-into-amino-acids" title="Permalink to this headline">¶</a></h2>
<p>How? Google translating DNA with BioPython.</p>
</div>
<div class="section" id="mafft-align-the-amino-acid-sequences">
<h2><strong>4.</strong> Mafft align the amino acid sequences<a class="headerlink" href="#mafft-align-the-amino-acid-sequences" title="Permalink to this headline">¶</a></h2>
<p>It’s the same syntax as step 1, just put your animo acids in a file.</p>
</div>
<div class="section" id="backtranslate-to-codons">
<h2><strong>5.</strong> Backtranslate to codons<a class="headerlink" href="#backtranslate-to-codons" title="Permalink to this headline">¶</a></h2>
<p>We need to backtranslate from amino acids into codons by inserting the gaps from the amino acid alignment into the nucleotide codons themselves.</p>
<p><strong>a.</strong> Go through each amino acid in the aligned amino acid array</p>
<p><strong>b.</strong> If it’s a regular amino acid, go ahead and take next three nucelotides from the degapped nucleotide array (the one with the n’s). If it’s a gap, think about how many gaps that corresponds to in nuceotide space (hint: it’s three) and insert those gaps.</p>
</div>
<div class="section" id="answer-key">
<h2>Answer Key<a class="headerlink" href="#answer-key" title="Permalink to this headline">¶</a></h2>
<p>Verify your answers:</p>
<p>The code itself can be found at <code class="docutils literal notranslate"><span class="pre">/srv/WI20/codon_align.py</span></code>.</p>
<p>The answers for each question are currently being made. They will be located in <code class="docutils literal notranslate"><span class="pre">/srv/WI20</span></code> (soon).</p>
<div class="section" id="congratulations-you-ve-completed-week-5-of-the-bioinformatics-crash-course">
<h3>Congratulations! You’ve completed Week 5 of the Bioinformatics Crash Course.<a class="headerlink" href="#congratulations-you-ve-completed-week-5-of-the-bioinformatics-crash-course" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>